### PR TITLE
Use discovery to test apply all status

### LINKED ISSUE #88901: Server-side Apply: Ownership not tracked for status subresource
**What happened**:

I've performed server-side apply on the `/status` subresource. e.g. 

```PATCH https://127.0.0.1:32768/apis/kubecarrier.io/v1alpha1/namespaces/default/<<crd name>>/test/status?fieldManager=name```

The request body is the following for example:

```json
{"apiVersion":"...","kind":"...","metadata":{"name":"test","namespace":"default","uid":"be8657c0-4391-4e45-8adb-6e7f3486f23a"},"status":{"conditions":[{"lastTransitionTime":"2020-03-06T14:09:19Z","message":"82","reason":"NewNumberGenerated","status":"True","type":"Established"}]}}
```

The CRD itself has status commonly defined as:
```go
type CRDStatus struct {
       ...
	// +listType=map
	// +listMapKey=type
	Conditions []CRDCondition `json:"conditions,omitempty"`
}
```

in the resulting openAPIv3 validation the correct object annotations are set for listType & listMapKey (thanks to controler-gen v0.2.5) 

**What you expected to happen**:

The patch is applied to the status subresource, and the `metadata.managedFields` is properly set. However, one of two things happens:

* if the status subfield conditions don't exist the request passes through and properly sets the `status.conditions` field. But the `metadata.managedFields` isn't set
* if the specific status condition type exits the request fails with 409 Conflict. It conflicts with other managers, e.g. `before-first-apply` since the `metadata.managedFields` is empty. 

if the `metadata.managedFields` exist beforehand for the status subresource, they are cleared after server-side apply on the `/status` subresource and become empty. 

**Anything else we need to know?**:

This is similar to https://github.com/kubernetes/kubernetes/issues/82046

Also somewhat related to https://github.com/kubernetes/kubernetes/issues/75564

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-11T18:14:22Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2020-01-14T00:09:19Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by apelisse**
Create a test that uses the discovery to apply status to all types of objects, and makes sure that it works everywhere.

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #88901

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/assign @liggitt

### CODE REVIEW

File: test/integration/apiserver/apply/status_test.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+const testNamespace = "statusnamespace"
+
+var statusData = map[schema.GroupVersionResource]string{
+	gvr("", "v1", "persistentvolumes"):                   `{"status": {"message": "hello"}}`,
+	gvr("", "v1", "resourcequotas"):                      `{"status": {"used": {"cpu": "5M"}}}`,
+	gvr("", "v1", "services"):                            `{"status": {"loadBalancer": {"ingress": [{"ip": "127.0.0.1"}]}}}`,
+	gvr("extension", "v1beta1", "ingresses"):             `{"status": {"loadBalancer": {"ingress": [{"ip": "127.0.0.1"}]}}}`,
+	gvr("autoscaling", "v1", "horizontalpodautoscalers"): `{"status": {"currentReplicas": 5}}`,
+}
+var statusDefault = `{"status": {"conditions": [{"type": "MyStatus", "status":"true"}]}}`
```
Comment (neolit123) [inline test/integration/apiserver/apply/status_test.go]: can be a `const`

File: test/integration/apiserver/apply/status_test.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+func hasStatus(resourceList *metav1.APIResourceList) bool {
+	for _, resource := range resourceList.APIResources {
+		if strings.HasSuffix(resource.Name, "/status") {
+			return true
+		}
+	}
+	return false
+}
+
+func findResource(resourceList *metav1.APIResourceList) *metav1.APIResource {
```
Comment (neolit123) [inline test/integration/apiserver/apply/status_test.go]: seems unused.

File: test/integration/apiserver/apply/status_test.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	gvr("extension", "v1beta1", "ingresses"):             `{"status": {"loadBalancer": {"ingress": [{"ip": "127.0.0.1"}]}}}`,
+	gvr("autoscaling", "v1", "horizontalpodautoscalers"): `{"status": {"currentReplicas": 5}}`,
+}
+var statusDefault = `{"status": {"conditions": [{"type": "MyStatus", "status":"true"}]}}`
+
+func gvr(g, v, r string) schema.GroupVersionResource {
+	return schema.GroupVersionResource{Group: g, Version: v, Resource: r}
+}
+
+func hasStatus(resourceList *metav1.APIResourceList) bool {
```
Comment (neolit123) [inline test/integration/apiserver/apply/status_test.go]: ditto.

File: pkg/registry/flowcontrol/flowschema/strategy.go [hunk 4]
```diff
@@ -94,7 +94,10 @@ var StatusStrategy = flowSchemaStatusStrategy{Strategy}
 func (flowSchemaStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
 	newFlowSchema := obj.(*flowcontrol.FlowSchema)
 	oldFlowSchema := old.(*flowcontrol.FlowSchema)
+
+	managedFields := newFlowSchema.ManagedFields
 	newFlowSchema.ObjectMeta = oldFlowSchema.ObjectMeta
+	newFlowSchema.ManagedFields = managedFields
```
Comment (liggitt) [inline pkg/registry/flowcontrol/flowschema/strategy.go]: this is delicate enough that this needs a comment explaining what's going on (same for other places)

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go [hunk 5]
```diff
@@ -252,7 +252,6 @@ func ResetObjectMetaForStatus(meta, existingMeta Object) {
 	meta.SetAnnotations(existingMeta.GetAnnotations())
 	meta.SetFinalizers(existingMeta.GetFinalizers())
 	meta.SetOwnerReferences(existingMeta.GetOwnerReferences())
-	meta.SetManagedFields(existingMeta.GetManagedFields())
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go]: suggest a placeholder comment indicating this omission is intentional, and why

File: test/integration/apiserver/apply/status_test.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (liggitt) [inline test/integration/apiserver/apply/status_test.go]: nit: 2020

File: test/integration/apiserver/apply/status_test.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+	gvr("", "v1", "resourcequotas"):                                     `{"status": {"used": {"cpu": "5M"}}}`,
+	gvr("", "v1", "services"):                                           `{"status": {"loadBalancer": {"ingress": [{"ip": "127.0.0.1"}]}}}`,
+	gvr("extensions", "v1beta1", "ingresses"):                           `{"status": {"loadBalancer": {"ingress": [{"ip": "127.0.0.1"}]}}}`,
+	gvr("networking.k8s.io", "v1beta1", "ingresses"):                    `{"status": {"loadBalancer": {"ingress": [{"ip": "127.0.0.1"}]}}}`,
+	gvr("autoscaling", "v1", "horizontalpodautoscalers"):                `{"status": {"currentReplicas": 5}}`,
+	gvr("batch", "v1beta1", "cronjobs"):                                 `{"status": {"lastScheduleTime": null}}`,
+	gvr("batch", "v2alpha1", "cronjobs"):                                `{"status": {"lastScheduleTime": null}}`,
+	gvr("storage.k8s.io", "v1", "volumeattachments"):                    `{"status": {"attached": true}}`,
+	gvr("policy", "v1beta1", "poddisruptionbudgets"):                    `{"status": {"currentHealthy": 5}}`,
+	gvr("certificates.k8s.io", "v1beta1", "certificatesigningrequests"): `{"status": {"conditions": [{"type": "Approved"}]}}`,
```
Comment (liggitt) [inline test/integration/apiserver/apply/status_test.go]: CSR status is weird... conditions don't have a status value (presence indicates "True") and approval conditions can only be set via the `/approval` subresource... suggest `{"status": {"conditions": [{"type": "MyStatus"}]}}`

File: test/integration/apiserver/apply/status_test.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+const statusDefault = `{"status": {"conditions": [{"type": "MyStatus", "status":"true"}]}}`
+
+var ignoreList = map[schema.GroupVersionResource]struct{}{
```
Comment (liggitt) [inline test/integration/apiserver/apply/status_test.go]: needs a comment about why these are ignored, and a warning to not add anything new to the list, and a TODO to fix/remove these

File: pkg/registry/flowcontrol/flowschema/strategy.go (hunk 4 above)
Comment (liggitt) [reply pkg/registry/flowcontrol/flowschema/strategy.go]: do we have any tests that ensure an API caller cannot mess with managedFields directly (i.e. that the managedFields in the new object at this point is computed by the field manager based on the old object's managedFields and `diff(old,new)`)?

File: test/integration/apiserver/apply/status_test.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+					t.Fatal(err)
+				}
+
+				statusObj := unstructured.Unstructured{}
+				if err := json.Unmarshal([]byte(status), &statusObj.Object); err != nil {
+					t.Fatal(err)
+				}
+				statusObj.SetAPIVersion(mapping.GroupVersionKind.GroupVersion().String())
+				statusObj.SetKind(mapping.GroupVersionKind.Kind)
+				statusObj.SetManagedFields(nil)
```
Comment (liggitt) [inline test/integration/apiserver/apply/status_test.go]: probably worth a comment about why this is needed

File: pkg/registry/flowcontrol/flowschema/strategy.go (hunk 4 above)
Comment (apelisse) [reply pkg/registry/flowcontrol/flowschema/strategy.go]: As discussed, we don't, this is a known limitation. We should document that, and also try to provide a fix asap.

**Comment by apelisse** [comment]
@liggitt Addressed feedback, PTAL!

File: pkg/registry/flowcontrol/flowschema/strategy.go (hunk 4 above)
Comment (riking) [reply pkg/registry/flowcontrol/flowschema/strategy.go:102]: Belated code review - this pattern would be a lot more clear if spelled as `temp := newObject.ObjectMeta.ManagedFields`.

**Pull request merged**